### DISCUSSION TITLE
Support to group endpoints into services

### DISCUSSION THREAD

**Discussion by alvarolopez** [upptime/upptime#121] [Ideas]
For complex setups, it would be desirable to have a way to group different endpoints into higher level services).

For instance, one could have a group "Webservers" with several endpoints, and another group "Database" with other endpoints, and so on and so forth. It should also be possible to generate downtimes, issues and interventions also to these services.

**Comment by AnandChowdhary** [comment]
Makes a lot of sense, yes! How do you envision this working? Say you add 3 sites to a "group", we will not only tell you the response time etc. of each site, but also a group average? And perhaps a collapsable in the README?

**Reply by Andre601** [reply]
What I feel like would be a good (visual) solution:

- On the site, have the grouped endpoints in their own "box". It would basically be a header with the group name and those entries in one area.
- On the readme, just split up by headers and have separate tables for these entries. Like one table for all in group A, one for all in group B, etc.

I hope you get what I imagine here...

**Comment by alvarolopez** [comment]
The most simple approach (that would work for us) is to implement them just as containers: i.e. only some way of grouping endpoints together, with a small description linked with an overall status: OK or DEGRADED. The endpoints will still have their own metrics, and the group can be collapsed or not.

For instance, this is how it looks in our current status page (that we are migrating tp upptime): https://status.ifca.es/

**Reply by josephtaylor-mn** [reply]
What do you mean by containers? How did you get this to work?

**Reply by Soneji** [reply]
@alvarolopez Hi
I also like this idea and think your proposed methodology of grouping is amazing! I would love to see this implemented

**Comment by wolfier74** [comment]
Add that pls

**Comment by Soneji** [comment]
One alternative way I thought this could be set up would be to have a keyword in the yaml configuration file, such as:

```yaml
sites:
  - name: example.com
    url: example.com
  - hr
  - name: example.com
    url: example.com
```
which could add a simple horizontal rule to the status website between monitors. 

Let me know your thoughts. 

PS Thank you for this great tool!

**Reply by AnandChowdhary** [reply]
Interesting, I was thinking something like:

```yaml
groups:
  - id: 1
    name: API endpoints
sites:
  - name: example.com
    url: example.com
    group-id: 1
  - name: example.com
    url: example.com
    group-id: 1
```

**Reply by Soneji** [reply]
This looks good! 👍

Maybe we could have groups insides the sites block and inside the group block have the corresponding sites

This would avoid an extra field in every site 

Just an idea, bit unsure of the yaml  

additionally I was thinking it would be cool to have an option as to which groups are expanded or not expanded by default in the status page, for which I think your original implementation would be better

**Reply by ririko5834** [reply]
Yes, you can add something like cstate has
Config for groups:
```
groups:
    - name: Group 1
      description: Example group 1

```
and then 
```
sites:
  - name: example.com
    url: example.com
    group: "Group 1"
```

The group description should be visible, when you place mouse on it:
![image](https://user-images.githubusercontent.com/64729293/104290786-0e45b900-54bb-11eb-80c3-6f6a8ac849ee.png)

**Reply by Soneji** [reply]
I think a numerical unique group-id value would be better than string matching

**Reply by ririko5834** [reply]
Ok, but add also group description

**Comment by nijel** [comment]
It would be also great to have an option to collapse a group by default -  this is useful to hide endpoints which are not interesting for most of the users. We currently use this with Cachet (https://status.weblate.org/).

**Comment by mai0li** [comment]
Was this somehow implemented? Don't want to create a duplicate discussion thread.

**Comment by PersonalDom** [comment]
Just found this thread after making my own *Facepalm* is their an update for this?

**Reply by PersonalDom** [reply]
I'll link my thread here, so you can take a look at how i've designed the yml and actual page (just as ideas.) 

[My thread.](https://github.com/upptime/upptime/discussions/385)

**Comment by LukePrior** [comment]
This sounds like a great feature to have.

**Comment by JuniYadi** [comment]
this is good if it can be made by group or category.

**Comment by JonathanTreffler** [comment]
This is not a solution for everybody here, as it does not allow for a parents status to be determined by its children uptime monitors, but if (like me) all you want is a visual way to group your monitors, you can achieve this using custom CSS:

```yml
status-website:
  css: "#sapper > main > section.live-status > article:nth-child(2), #sapper > main > section.live-status > article:nth-child(n+4):nth-child(-n+10) { margin-left: 40px; }"
```

The code above moves monitor 2 and monitors 4 -10  40px to the right, therefore making them appear grouped:
![image](https://user-images.githubusercontent.com/28999431/148657322-6461fa8c-69d6-4d6e-b5ee-f81c24618e61.png)
(in the screenshot monitors 3 - 6 are visible)

This is useful if you can determine the health of a server independently from knowing if all services on it are running.

**Comment by GammaPi** [comment]
I prefer the maintainer's approach:

```
groups:
  - id: 1
    name: API endpoints
sites:
  - name: example.com
    url: example.com
    group-id: 1
  - name: example.com
    url: example.com
    group-id: 1
```

**Reply by GammaPi** [reply]
Basically it means adding group-id in https://github.com/upptime/uptime-monitor/blob/66bdb6b30608c39de16767318b96085705596084/src/summary.ts#L56.

And add some logic here https://github.com/upptime/status-page/blob/d85d1c8226a71e95925e25bdd7db96e915cfdf42/src/components/LiveStatus.svelte#L100

**Reply by GammaPi** [reply]
Does h4 look good?
![image](https://user-images.githubusercontent.com/19838874/210158169-0c6b87b4-41dc-47d4-ab24-a2acd85680bf.png)

**Reply by AnandChowdhary** [reply]
Yes, I think we can add some margins above the heading for better visual separation, but this is exactly how I envisioned it. Thanks for your hard work!

**Reply by alvarolopez** [reply]
So cool, is there any chance this gets implemented soon?

**Reply by AnandChowdhary** [reply]
I'm happy to accept a PR, would you be up for it? I can also guide you on what to do! :)

**Reply by Tealk** [reply]
any news on the proposal?

**Reply by goncalopolido** [reply]
Is there any ETA for this feature? It would be incredibly useful for organizing larger setups.
